### PR TITLE
Update for 2019

### PR DESCRIPTION
**Pull request by selbekk**
This pull request upgrades a lot of dependencies, and rewrites the code to avoid having a custom server.

The TL;DR is:

- Fetch articles via the `frontmatter-markdown-loader` instead over the API
- Move files from `/static` to `/public` (deprecation notice)
- Remove `prefetch` flags on links (next does this based on heuristics now)
- Use `babel-plugin-root-import` to avoid `../../../../..` imports (particularly visible with Next's new file system based dynamic route architecture
- Rewrite routes to use Next's new file system based dynamic route architecture.
- Remove the server altogether

There's a few regressions you need to be aware of:
- The `/today` route now redirects to today's route. This can be fixed if we're bothered to refactor a bit
- We no longer use `helmet` for setting security related headers. I don't see a way we could keep this without re-adding a custom server. This should honestly be a part of Next itself, perhaps we can file a pull request? 
- The transition between posts is no longer flawless. This should probably be adressed.

### CODE REVIEW

**Comment by selbekk** [comment]
Created a [feature request](https://github.com/zeit/next.js/issues/9013) issue on the Next repo

File: components/article-list.js [hunk 1]
```diff
@@ -89,12 +89,9 @@ const ArticleList = props => {
         return (
           <Leaf key={day} index={day}>
             <Link
-              href={`/post?year=${props.year}&date=${day}${
-                hackerMode ? '&mode=hacker' : ''
-              }`}
+              href={`/[year]/[date]${hackerMode ? '?mode=hacker' : ''}`}
```
Comment (sveinpg) [inline components/article-list.js:92]: @selbekk Hvordan syntax er dette?

Comment (selbekk) [reply components/article-list.js:92]: Dynamisk routing! https://nextjs.org/docs#dynamic-routing

Comment (selbekk) [reply components/article-list.js:92]: Basically så har jeg renamet `post.js` til `[date].js` og plassert den inni mappen `[year]`. Med klammer og det hele. Det er sånn next nå støtter filsystembasert dynamisk routing.

File: components/site-footer.js [hunk 2]
```diff
@@ -85,28 +89,26 @@ const SiteFooter = () => {
...
 
         <Column>
           <Heading>Proudly made by</Heading>
-          <Link href="https://www.bekk.no">
-            <a aria-label="Bekk">
-              <BekkLogo style={{ maxWidth: '100px' }} />
-            </a>
-          </Link>
+          <WhiteExternalLink href="https://www.bekk.no" aria-label="Bekk">
+            <BekkLogo style={{ maxWidth: '100px' }} />
```
Comment (sveinpg) [inline components/site-footer.js]: Når man først bruker styled componets kan man bruke css-prop på inline styling: https://www.styled-components.com/docs/api#css-prop

File: components/site-header.js [hunk 3]
```diff
@@ -33,9 +33,10 @@ const LinkText = styled.a`
 const SiteHeader = () => (
   <Container>
     <SiteTitle>
-      <Link prefetch href="/" passHref>
+      <Link href="/" passHref>
         <LinkText>
-          {siteConfig.name.split('.')[0]} <ChristmasTree style={{ width: '1.5em' }} /> christmas
+          {siteConfig.name.split('.')[0]}{' '}
```
Comment (sveinpg) [inline components/site-header.js:38]: Personlig liker jeg ikke `{' '}` så godt 🤷‍♂

Comment (selbekk) [reply components/site-header.js:38]: Personlig bruker jeg `prettier` ;-) Enig i at det ikke ser så fint ut, men meh - det gjøres automatisk.

File: components/site-footer.js (hunk 2 above)
Comment (selbekk) [reply components/site-footer.js]: Kult, det kan jeg sikkert fikse :)

File: components/site-header.js [hunk 4]
```diff
@@ -33,9 +33,10 @@ const LinkText = styled.a`
 const SiteHeader = () => (
   <Container>
     <SiteTitle>
-      <Link prefetch href="/" passHref>
```
Comment (sveinpg) [inline components/site-header.js:36]: Dette har blitt default i v9?

Comment (selbekk) [reply components/site-header.js:36]: Jepp, den bruker `IntersectionObserver` til å preloade alle lenker som er i viewporten. Man kan opte ut av det hvis man vil med `prefetch={false}`

File: components/typography.js [hunk 5]
```diff
@@ -57,8 +57,12 @@ const StyledLink = styled.a`
   cursor: pointer;
   text-decoration: underline;
 `;
-export const LinkText = ({ href, as, prefetch, ...rest }) => (
-  <Link href={href} as={as} prefetch={prefetch} passHref={true}>
+export const LinkText = ({ href, as, ...rest }) => (
+  <Link href={href} as={as} passHref={true}>
```
Comment (sveinpg) [inline components/typography.js:61]: Hva gjør `passHref` propen?

File: pages/[year].js [hunk 6]
```diff
@@ -54,8 +52,23 @@ const YearPage = props => {
   );
 };
 
-YearPage.getInitialProps = async context => ({
-  year: Number(context.query.year)
-});
+YearPage.getInitialProps = async ({ query, res }) => {
+  if (/^\d{1,2}$/.test(query.year)) {
```
Comment (sveinpg) [inline pages/[year].js:56]: Hva tester du her?

File: components/typography.js (hunk 5 above)
Comment (selbekk) [reply components/typography.js:61]: > If child is an `<a>` tag and doesn't have a href attribute we specify it so that the repetition is not needed by the user. However, sometimes, you’ll want to pass an `<a>` tag inside of a wrapper and the Link won’t recognize it as a hyperlink, and, consequently, won’t transfer its href to the child. In cases like that, you should define a boolean `passHref` property to the Link, forcing it to expose its href property to the child.

Siden barnet her ikke er en ren `<a />`, men en `styled.a`, så må jeg sette `passHref={true}`

Docsa her er ganske gode: https://nextjs.org/docs#forcing-the-link-to-expose-href-to-its-child

File: components/article-list.js (hunk 1 above)
Comment (sveinpg) [reply components/article-list.js:92]: Stilig

File: pages/[year].js (hunk 6 above)
Comment (selbekk) [reply pages/[year].js:56]: jeg sjekker om urlen er `/{ett eller to tall}`, som `/1`, `/12` osv. Dette var URLene jeg brukte første året for `react.christmas`, før jeg flyttet over til `/2017/1`, `/2017/12` osv. :) Ref kommentaren under.

**Review by sveinpg — APPROVED** [review]

File: pages/[year].js (hunk 6 above)
Comment (sveinpg) [reply pages/[year].js:56]: Grunnen til at jeg spurte var at det ikke kom helt tydelig frem i kommentaren
